lib/spack/spack/new_installer.py

```diff
@@ -0,0 +1,1194 @@
+                        self.build_status.next(-1)
+
+                # Flush installed packages to the database and enqueue any parents that are now
+                # ready.
+                if to_insert_in_database and self._save_to_db(to_insert_in_database):
+                    for entry in to_insert_in_database:
+                        self._enqueue_parents(entry.spec.dag_hash())
+                    to_insert_in_database.clear()
+
+                # Again, the first job should start immediately and does not require a token.
+                if self.pending_builds and not self.running_builds:
+                    self._start(selector, jobserver)
+
+                # For the rest we try to obtain tokens from the jobserver.
+                if self.pending_builds and jobserver_token_available:
+                    # Then we try to schedule as many jobs as we can acquire tokens for.
+                    max_new_jobs = len(self.pending_builds)
+                    for _ in range(jobserver.acquire(max_new_jobs)):
+                        self._start(selector, jobserver)
+
+                # Finally update the UI
+                self.build_status.update()
+        except KeyboardInterrupt:
+            # Cleanup running builds.
+            for child in self.running_builds.values():
+                child.proc.terminate()
+            for child in self.running_builds.values():
+                child.proc.join()
+            for child in self.running_builds.values():
+                shutil.rmtree(child.spec.prefix, ignore_errors=True)
+            raise
+        finally:
+            # Restore terminal settings
+            if old_stdin_settings:
+                termios.tcsetattr(sys.stdin, termios.TCSADRAIN, old_stdin_settings)
+
+            # Clean up resources
+            # Final cleanup of any remaining finished packages before exit
+            self.build_status.overview_mode = True
+            self.build_status.update(finalize=True)
+            selector.close()
+            jobserver.close()
+            old_wakeup_fd = signal.set_wakeup_fd(-1)
+            os.close(old_wakeup_fd)
+            os.close(signal_r)
+
+        if failures:
+            lines = [f"{s}: {s.package.log_path}" for s in failures]
+            raise spack.error.InstallError(
+                "The following packages failed to install:\n" + "\n".join(lines)
+            )
+
+    def _save_to_db(self, to_insert_in_database: List[ChildInfo]) -> bool:
+        db = spack.store.STORE.db
+        try:
+            # Only try to get the lock once (non-blocking). If it fails, try it next time.
+            if db.lock.acquire_write(timeout=1e-9):
+                db._read()
+        except spack.util.lock.LockTimeoutError:
+            return False
+        try:
+            for entry in to_insert_in_database:
+                db._add(entry.spec, explicit=entry.explicit)
+            return True
+        finally:
+            db.lock.release_write(db._write)
```

Comment:
I think this could be accomplished more cleanly by adding a `timeout` argument to the `Database.write_transaction` method. That argument would be passed through to the constructor for `Database._write_transaction_impl` which is a `spack.llnl.util.LockTransaction` object.

Then this can be simplified to

```suggestion
    def _save_to_db(self, to_insert_in_database: List[ChildInfo]) -> bool:
        db = spack.store.STORE.db
        try:
            # Only try to get the lock once (non-blocking). If it fails, try it next time.
            with db.write_transaction(timeout=1e-9):
                for entry in to_insert_in_database:
                    db._add(entry.spec, explicit=entry.explicit)
                return True
        except spack.util.lock.LockTimeoutError:
            return False
```

Reply:
I agree, and the `1e-9` constant could be removed with `blocking=False` or something. However, that would extend the scope of the PR to database.py, which I want to avoid.

Reply:
I'm fine with making that future work, just don't want to lose track of it

Reply:
See #51434